agent_build_refactored/managed_packages/managed_packages_builders.py

```diff
@@ -0,0 +1,932 @@
+            if package_path.name not in packages_to_publish:
+                logger.info(f"Package {package_path.name} has been skipped since it's already in repo.")
+                continue
+
+            repo_paths = {
+                "deb": f"{user_name}/{repo_name}/any/any",
+                "rpm": f"{user_name}/{repo_name}/rpm_any/rpm_any",
+            }
+
+            check_call_with_log(
+                [
+                    "package_cloud",
+                    "push",
+                    repo_paths[self.PACKAGE_TYPE],
+                    str(package_path)
+                ]
+            )
+
+            logging.info(f"Package {package_path.name} is published.")
+
+    def _search_for_packages_in_repo(
+            self,
+            params: Dict,
+            token: str,
+            user_name: str,
+            repo_name: str,
+    ) -> List:
+        """
+        Search packages with the given query params in the Packagecloud repository.
+        :param params: Params for the url query.
+        :param token: Packagecloud token
+        :param repo_name: Target Packagecloud repo.
+        :param user_name: Target Packagecloud user.
+        :return: List of found packages.
+        """
+        import requests
+        from requests.auth import HTTPBasicAuth
+
+        auth = HTTPBasicAuth(token, "")
+
+        param_filters = {
+            "deb": "debs",
+            "rpm": "rpms"
+        }
+
+        params["filter"] = param_filters[self.PACKAGE_TYPE]
+
+        with requests.Session() as s:
+            resp = s.get(
+                url=f"https://packagecloud.io/api/v1/repos/{user_name}/{repo_name}/search.json",
+                params=params,
+                auth=auth
+            )
+
+        resp.raise_for_status()
+
+        packages = resp.json()
+
+        # filter only packages with appropriate architecture.
+
+        arch_packages = [p for p in packages]
+        return arch_packages
+
+    def find_last_repo_package(
+            self,
+            package_name: str,
+            token: str,
+            user_name: str,
+            repo_name: str,
+
+    ) -> Optional[str]:
+        """
+        Find the most recent version of the given package in the repo.
+        :param package_name: Name of the package.
+        :param token: Packagecloud token
+        :param repo_name: Target Packagecloud repo.
+        :param user_name: Target Packagecloud user.
+        :return: filename of the package if found, or None.
+        """
+
+        packages = self._search_for_packages_in_repo(
+            params={
+                "q": f"{package_name}",
+            },
+            token=token,
+            user_name=user_name,
+            repo_name=repo_name
+        )
+
+        if len(packages) == 0:
+            logger.info(f"Could not find any package with name {package_name}")
+            return None
+
+        last_package = packages[-1]
```

Comment:
So search.json API has guaranted and stable sort order? If so, what is based on? package id / upload timestamp or similar?

Reply:
It has to group by distro_version  (aka `Ubuntu/22.04`) and then sort by the package version inside each group. 
But as it was said in previous comment  https://github.com/scalyr/scalyr-agent-2/pull/1025#discussion_r1030641030, new do not rely on that now and do full run through all pages.